### PR TITLE
[MM-58017] Calls usage stats

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR implements some basic usage stats for Calls to be rendered in the System Statistics page (`/admin_console/reporting/system_analytics`), namely:

- Total Calls
- Total Active Calls (i.e., ongoing)
- Total Active Sessions
- Calls by day
- Calls by month
- Calls by channel type
- Average call duration
- Average call participants

Note: the last commit is purely about fixing type errors after updating the web app dependency.

#### Screenshots

image
image
image

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-58017

### CODE REVIEW

File: server/db/stats_store.go
Comment (streamer45) [inline server/db/stats_store.go:117]: Keeping it simple for now, but ideally, most of these methods could filter for the last month or so. That way they would perform much better.

File: server/db/stats_store.go
Comment (streamer45) [inline server/db/stats_store.go:204]: To make this more performant we could add an index on `calls.StartAt`.

File: server/api.go
Comment (cpoile) [inline server/api.go:544]: Would you consider making it a ticket? (or did you mean to clean this up before submitting?)

File: server/db/stats_store.go
Comment (cpoile) [inline server/db/stats_store.go:268]: I don't really mind, but does it matter (30 vs 28 vs 31)?

File: server/api_router.go
Comment (cpoile) [inline server/api_router.go:152]: Considering the weight of the call, should we at least rate limit this pretty heavily?

File: webapp/src/index.tsx
Comment (cpoile) [inline webapp/src/index.tsx:435]: Would it be cleaner to put this fn in its own file? I only ask because it seems a little out of place considering the conciseness of the other register calls above.

**Review by cpoile — COMMENTED** [review]
Just minor things, looks great!

File: server/api.go
Comment (streamer45) [reply server/api.go:544]: Yes, I can make a ticket. I just don't know yet if this is really needed. I almost feel it may be best to go with https://github.com/mattermost/mattermost-plugin-calls/pull/762#discussion_r1626483383.

File: server/db/stats_store.go
Comment (streamer45) [reply server/db/stats_store.go:268]: Tracking the last 30 days seemed more straightforward than having to deal with month-by-month cases.

File: server/api_router.go
Comment (streamer45) [reply server/api_router.go:152]: We could, but the global rate limiter we have on API endpoints would already apply to this admin-only endpoint coming from a pretty specific page, so I am not sure if the extra complication is worth it.

File: webapp/src/index.tsx
Comment (streamer45) [reply webapp/src/index.tsx:435]: Yes, I was planning to do it but then I couldn't think of a great place. I can add it to utils I suppose.

**Comment by streamer45** [comment]
@cpoile Let me know if there's anything else blocking.

**Review by cpoile — APPROVED** [review]
Awesome, looks great!

**Comment by streamer45** [comment]
@DHaussermann This is the matching PR you can use to test https://github.com/mattermost/mattermost/pull/27188 if you wish.

**Comment by DHaussermann** [comment]
@streamer45 I have this deployed and took a look ver things. I'm a bit confused by what should be visible and testable right now. 

The data I **can** see appears accurate after creating a few data points but... I can't see the following panels

- Calls by day
- Calls by month
- Calls by channel type (donut graph)
- Daily Calls (line graph)
- Monthly Calls (line graph)

0/5 Perhaps these are being hidden because it's a new server and I don't have data points (though I'd still want the donut graph of calls per channel type)

**Comment by streamer45** [comment]
> @streamer45 I have this deployed and took a look ver things. I'm a bit confused by what should be visible and testable right now.
> 
> The data I **can** see appears accurate after creating a few data points but... I can't see the following panels
> 
> * Calls by day
> * Calls by month
> * Calls by channel type (donut graph)
> * Daily Calls (line graph)
> * Monthly Calls (line graph)
> 
> 0/5 Perhaps these are being hidden because it's a new server and I don't have data points (though I'd still want the donut graph of calls per channel type)

@DHaussermann That's likely because of the version check. Those would only display on v9.10 so you may need to check on a release candidate or something. We can defer to later if it helps.

**Comment by DHaussermann** [comment]
Okay thanks. I was wondering that. I tried to work around this be deploying on `master` but then the minimum version is not met :)  

Yes - it would be reasonable to test the remaining panels post merge in an RC. 👍

**Comment by streamer45** [comment]
@DHaussermann Will go ahead and prepare this to be merged then to unblock the release. We can come back to this again once we have an RC.

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
I built some data for this and then observed the stats. The figures shown seem accurate based on the Calls usage eon the server for the following:
- Total Calls
- Total Active Calls
- Total Active Sessions
- Avg Call Duration (minutes)
- Avg Call Participants

LGTM!

As mentioned above, I will test the remaining panels post merge.

**Comment by ThiefMaster** [comment]
Is "Calls by channel type" only available in EE by any chance? If yes, that feels rather strange considering that with #809 you paywall a feature that people may currently be using, but at the same time make it impossible for admins of the free version to actually see if and how much it is being used (which may actually be information that could encourage someone to buy a license!)...

**Comment by streamer45** [comment]
> Is "Calls by channel type" only available in EE by any chance? If yes, that feels rather strange considering that with #809 you paywall a feature that people may currently be using, but at the same time make it impossible for admins of the free version to actually see if and how much it is being used (which may actually be information that could encourage someone to buy a license!)...

@ThiefMaster They require a newer MM server version to be rendered (v9.10+ I believe) and not just plugin v0.28+. You can always check whether `http://localhost:8065/plugins/com.mattermost.calls/stats` returns something.

**Comment by ThiefMaster** [comment]
Ah, good point. I'm still on v9.9.x. Anyway, I spotted the stats endpoint in the meantime and got the information from there :)

**Pull request merged**
